### PR TITLE
fix: duplicate entry exception for correlation_ids table.

### PR DESCRIPTION
**Pull request by armory-abedonik**

### CODE REVIEW

**Comment by dbyron-sf** [comment]
tests?

**Comment by armory-abedonik** [comment]
@dbyron-sf , we already have a few tests that covers these changes. For instance this one:
https://github.com/spinnaker/orca/blob/8050e9edd058bae52506bb9e5dcc06267f21b976/orca-core-tck/src/main/groovy/com/netflix/spinnaker/orca/pipeline/persistence/PipelineExecutionRepositoryTck.groovy#L553

Thank you!

**Comment by dbyron-sf** [comment]
@armory-abedonik Thanks for adding another test.  Looks like a test is failing...not the new one and not the one you mentioned.

File: orca-redis/src/main/java/com/netflix/spinnaker/orca/pipeline/persistence/jedis/RedisExecutionRepository.java [hunk 1]
```diff
@@ -695,7 +695,7 @@ public PipelineExecution retrieveByCorrelationId(
   @Override
   public PipelineExecution retrievePipelineForCorrelationId(@Nonnull String correlationId)
       throws ExecutionNotFoundException {
-    String key = format("pipelineCorrelation:%s", correlationId);
+    String key = format("correlation:%s", correlationId);
```
Comment (dbyron-sf) [inline orca-redis/src/main/java/com/netflix/spinnaker/orca/pipeline/persistence/jedis/RedisExecutionRepository.java:698]: This seems like a pretty big change.  Why do we need it?

Comment (armory-abedonik) [reply orca-redis/src/main/java/com/netflix/spinnaker/orca/pipeline/persistence/jedis/RedisExecutionRepository.java:698]: In the absence of any existing code that generates keys predicated upon the prefix `pipelineCorrelation`, it follows that `pipelineCorrelation` will not be present within our database. Given that pipeline and orchestration executions are endowed with unique identifiers, we may opt to employ the same key prefix for both.

Comment (armory-abedonik) [reply orca-redis/src/main/java/com/netflix/spinnaker/orca/pipeline/persistence/jedis/RedisExecutionRepository.java:698]: https://github.com/spinnaker/orca/blob/6dfe7ddca68d4650eea3010b2736804be7a30cbd/orca-redis/src/main/java/com/netflix/spinnaker/orca/pipeline/persistence/jedis/RedisExecutionRepository.java#L1366

**Review by kkotula — APPROVED** [review]
LGTM

File: orca-core-tck/src/main/groovy/com/netflix/spinnaker/orca/pipeline/persistence/PipelineExecutionRepositoryTck.groovy [hunk 2]
```diff
@@ -550,6 +550,28 @@ abstract class PipelineExecutionRepositoryTck<T extends ExecutionRepository> ext
     status << ExecutionStatus.values()
   }
 
+  def "should return task ref for currently running pipeline by correlation id"() {
```
Comment (dbyron-sf) [inline orca-core-tck/src/main/groovy/com/netflix/spinnaker/orca/pipeline/persistence/PipelineExecutionRepositoryTck.groovy:553]: I see that this test nicely fails when I run
```
$ ./gradlew orca-redis:test --tests JedisPipelineExecutionRepositorySpec
```
without the redis fix, and passes with it.

But, neither
```
$ ./gradlew orca-sql:test --tests **MySqlPipelineExecutionRepositorySpec
```
nor
```
$ ./gradlew orca-sql:test --tests PgSqlPipelineExecutionRepositorySpec
```
fail without the SqlExecutionRepository change.  Is there some other test that demonstrates the bug?

Comment (armory-abedonik) [reply orca-core-tck/src/main/groovy/com/netflix/spinnaker/orca/pipeline/persistence/PipelineExecutionRepositoryTck.groovy:553]: The occurrence of the "duplicate entry exception for correlation_ids table" bug arises when two or more concurrent instances of the Orca service attempt to write identical "correlation id" simultaneously. Presently, all existing tests are executed sequentially within a single thread, thereby preventing the replication of this bug. However, it is essential to note that these tests primarily serve the purpose of verifying the correctness of the business logic.

Comment (armory-abedonik) [reply orca-core-tck/src/main/groovy/com/netflix/spinnaker/orca/pipeline/persistence/PipelineExecutionRepositoryTck.groovy:553]: I have reservations regarding our ability to construct a unit test capable of adequately cover this particular bug scenario. Even in the event that we endeavor to establish several distinct processes executed concurrently on a CPU with multiple cores, achieving simultaneous data write operations in 100% of cases appears unattainable.

Comment (dbyron-sf) [reply orca-core-tck/src/main/groovy/com/netflix/spinnaker/orca/pipeline/persistence/PipelineExecutionRepositoryTck.groovy:553]: It does seem like there'd be a way to demonstrate that the old code was busted by writing two pipelines with the same correlation id in sequence.  Granted, it's more difficult to verify that the new code is working correctly, and I agree throwing a bunch of threads at it hoping isn't great.  If there's a way to spy on `onConflict` or `doNothing`, we could potentially do it, but you've convinced me that we're ok without an explicit test for this.

**Review by dbyron-sf — APPROVED** [review]

**Comment by jasonmcintosh** [comment]
@Mergifyio backport release-1.32.x release-1.31.x release-1.30.x release-1.28.x

**Pull request merged**